src/utils/translator.rs

```diff
@@ -0,0 +1,494 @@
+        (map $arg:ident lanes) => (*$arg);
+
+        // This case takes the arguments of a wasmparser instruction and creates
+        // a wasm-encoder instruction. There are a few special cases for where
+        // the structure of a wasmparser instruction differs from that of
+        // wasm-encoder.
+        (build $op:ident) => (I::$op);
+        (build BrTable $arg:ident) => (I::BrTable($arg.0, $arg.1));
+        (build I32Const $arg:ident) => (I::I32Const(*$arg));
+        (build I64Const $arg:ident) => (I::I64Const(*$arg));
+        (build F32Const $arg:ident) => (I::F32Const(f32::from_bits($arg.bits())));
+        (build F64Const $arg:ident) => (I::F64Const(f64::from_bits($arg.bits())));
+        (build V128Const $arg:ident) => (I::V128Const($arg.i128()));
+        (build $op:ident $arg:ident) => (I::$op($arg));
+        (build CallIndirect $ty:ident $table:ident $_:ident) => (I::CallIndirect {
+            ty: $ty,
+            table: $table,
+        });
+        (build ReturnCallIndirect $ty:ident $table:ident) => (I::ReturnCallIndirect {
+            ty: $ty,
+            table: $table,
+        });
+        (build MemoryGrow $mem:ident $_:ident) => (I::MemoryGrow($mem));
+        (build MemorySize $mem:ident $_:ident) => (I::MemorySize($mem));
+        (build $op:ident $($arg:ident)*) => (I::$op { $($arg),* });
+    }
+
+	wasmparser::for_each_operator!(translate)
+}
+
+pub fn block_type(t: &mut dyn Translator, ty: &wasmparser::BlockType) -> Result<BlockType> {
+	match ty {
+		wasmparser::BlockType::Empty => Ok(BlockType::Empty),
+		wasmparser::BlockType::Type(ty) => Ok(BlockType::Result(t.translate_ty(ty)?)),
+		wasmparser::BlockType::FuncType(f) => Ok(BlockType::FunctionType(t.remap(Item::Type, *f)?)),
+	}
+}
+
+pub fn memarg(t: &mut dyn Translator, memarg: &wasmparser::MemArg) -> Result<MemArg> {
+	Ok(MemArg {
+		offset: memarg.offset,
+		align: memarg.align.into(),
+		memory_index: t.remap(Item::Memory, memarg.memory)?,
+	})
+}
+
+#[allow(dead_code)]
+pub fn data(t: &mut dyn Translator, data: wasmparser::Data<'_>, s: &mut DataSection) -> Result<()> {
+	let offset;
+	let mode = match &data.kind {
+		DataKind::Active { memory_index, offset_expr } => {
+			offset = t.translate_const_expr(
+				offset_expr,
+				&wasmparser::ValType::I32,
+				ConstExprKind::DataOffset,
+			)?;
+			DataSegmentMode::Active {
+				memory_index: t.remap(Item::Memory, *memory_index)?,
+				offset: &offset,
+			}
+		},
+		DataKind::Passive => DataSegmentMode::Passive,
+	};
+	s.segment(DataSegment { mode, data: data.data.iter().copied() });
+	Ok(())
+}
+
+pub fn code(t: &mut dyn Translator, body: FunctionBody<'_>, s: &mut CodeSection) -> Result<()> {
+	let locals = body
+		.get_locals_reader()?
+		.into_iter()
+		.map(|local| {
+			let (cnt, ty) = local?;
+			Ok((cnt, t.translate_ty(&ty)?))
+		})
+		.collect::<Result<Vec<_>>>()?;
+	let mut func = Function::new(locals);
+
+	let mut reader = body.get_operators_reader()?;
+	reader.allow_memarg64(true);
```

Comment:
Why `allow_memarg64`? Should be enabled only when memory64 feature is enabled.

Reply:
good point